client/web/web_test.go

```diff
@@ -62,3 +72,110 @@ func TestQnapAuthnURL(t *testing.T) {
 		})
 	}
 }
+
+// TestServeAPI tests the web client api's handling of
+//  1. invalid endpoint errors
+//  2. host name validation errors
+//  3. proxing to the localapi
+func TestServeAPI(t *testing.T) {
+	lal := memnet.Listen("local-tailscaled.sock:80")
+	defer lal.Close()
+
+	// Serve dummy localapi. Just returns "success".
+	go func() {
+		localapi := &http.Server{Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			fmt.Fprintf(w, "success")
+		})}
+		defer localapi.Close()
+		if err := localapi.Serve(lal); err != nil {
+			t.Error(err)
+		}
+	}()
+	s := &Server{lc: &tailscale.LocalClient{Dial: lal.Dial}}
+
+	dummyNodeHostInfo := tailcfg.Hostinfo{
+		Hostname: "node.local",
+	}
+	dummyNode := tailcfg.Node{
+		Name:      "node.tailnet.ts.net.",
+		Hostinfo:  dummyNodeHostInfo.View(),
+		Addresses: []netip.Prefix{netip.MustParsePrefix("100.101.102.103/32")},
+	}
+	attachDummyNode := func() { s.updateSelf(dummyNode.View()) }
+
+	tests := []struct {
```

Comment:
a couple more test suggestions:
* empty Host header
* malformed IP address like `1.2.3:100`
* IPv6 address
* "garbage string that's neither IP nor hostname"
* node's public IP? that should probably never get exposed, right?
* `localhost` without port number